test/unit/controllers/declaration/ItemTypeControllerSpec.scala

```diff
@@ -0,0 +1,245 @@
+
+        val result = controller.submitItemType(itemId)(postRequest(correctForm))
+
+        status(result) mustBe BAD_REQUEST
+        verify(mockItemTypePage, times(0)).apply(any(), any(), any(), any(), any())(any(), any())
+      }
+
+      "form action is incorrect" in {
+
+        withNewCaching(aDeclaration(withItem(anItem(withItemId(itemId)))))
+
+        val wrongAction = Seq(
+          ("combinedNomenclatureCode", "code"),
+          ("taricAdditionalCode[]", ""),
+          ("nationalAdditionalCode[]", ""),
+          ("descriptionOfGoods", "description"),
+          ("cusCode", ""),
+          ("unDangerousGoodsCode", ""),
+          ("statisticalValue", "value"),
+          ("WrongAction", "")
+        )
+
+        val result = controller.submitItemType(itemId)(postRequestAsFormUrlEncoded(wrongAction: _*))
+
+        status(result) mustBe BAD_REQUEST
+        verify(mockItemTypePage, times(0)).apply(any(), any(), any(), any(), any())(any(), any())
+      }
+
+      "form is incorrect during adding" in {
+
+        withNewCaching(aDeclaration(withItem(anItem(withItemId(itemId)))))
+
+        val incorrectForm = Seq(
+          ("combinedNomenclatureCode", "!@#$$%"),
+          ("taricAdditionalCode[]", "!@#$$%"),
+          ("nationalAdditionalCode[]", "!@#$$%"),
+          ("descriptionOfGoods", "!@#$$%"),
+          ("cusCode", "!@#$$%"),
+          ("unDangerousGoodsCode", "!@#$$%"),
+          ("statisticalValue", "!@#$$%"),
+          addActionUrlEncoded
+        )
+
+        val result = controller.submitItemType(itemId)(postRequestAsFormUrlEncoded(incorrectForm: _*))
+
+        status(result) mustBe BAD_REQUEST
+        verify(mockItemTypePage, times(1)).apply(any(), any(), any(), any(), any())(any(), any())
+      }
+
+      "form is incorrect during saving" in {
+
+        withNewCaching(aDeclaration(withItem(anItem(withItemId(itemId)))))
+
+        val incorrectForm = Seq(
+          ("combinedNomenclatureCode", ""),
+          ("taricAdditionalCode[]", "!@#$$%"),
+          ("nationalAdditionalCode[]", ""),
+          ("descriptionOfGoods", ""),
+          ("cusCode", ""),
+          ("unDangerousGoodsCode", ""),
+          ("statisticalValue", ""),
+          saveAndContinueActionUrlEncoded
+        )
+
+        val result = controller.submitItemType(itemId)(postRequestAsFormUrlEncoded(incorrectForm: _*))
+
+        status(result) mustBe BAD_REQUEST
+        verify(mockItemTypePage, times(1)).apply(any(), any(), any(), any(), any())(any(), any())
+      }
+    }
+
+    "return 303 (SEE_OTHER)" when {
+
+      "item doesn't exists in the cache" in {
+
+        withNewCaching(aDeclaration())
+
+        val result = controller.displayPage(itemId)(getRequest())
+
+        status(result) mustBe SEE_OTHER
+        verify(mockItemTypePage, times(0)).apply(any(), any(), any(), any(), any())(any(), any())
+      }
+
+      "correct item is added and declaration model exist in the cache" in {
```

Comment:
This is confusing, because in our case the `item` is something bigger. And you are actually not adding the whole `item`, but only some information about it. Instead of `item` I'd suggest even the simplest `data`.

Reply:
You're right, changing....